kion/resource_funding_source_permission_mapping.go

```diff
@@ -0,0 +1,240 @@
+				Description: "Set of user group IDs for the permission mapping (must be provided in numerical order).",
+			},
+			"user_ids": {
+				Type:        schema.TypeSet,
+				Required:    true,
+				Elem:        &schema.Schema{Type: schema.TypeInt},
+				Description: "Set of user IDs for the permission mapping (must be provided in numerical order).",
+			},
+		},
+	}
+}
+
+func resourceFundingSourcePermissionsMappingCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*hc.Client)
+
+	fundingSourceID := d.Get("funding_source_id").(int)
+	appRoleID := d.Get("app_role_id").(int)
+
+	userGroupsIDs := hc.ConvertInterfaceSliceToIntSlice(d.Get("user_groups_ids").(*schema.Set).List())
+	userIDs := hc.ConvertInterfaceSliceToIntSlice(d.Get("user_ids").(*schema.Set).List())
+
+	mapping := hc.FundingSourcePermissionsMapping{
+		AppRoleID:     appRoleID,
+		UserGroupsIDs: userGroupsIDs,
+		UserIDs:       userIDs,
+	}
+
+	err := client.PATCH(fmt.Sprintf("/v3/funding-source/%d/permission-mapping", fundingSourceID), []hc.FundingSourcePermissionsMapping{mapping})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fmt.Sprintf("%d-%d", fundingSourceID, appRoleID))
+
+	// Ensure the state reflects the provided list
+	if err := d.Set("user_groups_ids", userGroupsIDs); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("user_ids", userIDs); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceFundingSourcePermissionsMappingRead(ctx, d, m)
+}
+
+func resourceFundingSourcePermissionsMappingRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*hc.Client)
+
+	parts := strings.Split(d.Id(), "-")
+	if len(parts) != 2 {
+		return diag.Errorf("invalid resource ID format, expected {funding_source_id}-{app_role_id}")
+	}
+
+	fundingSourceID, err := strconv.Atoi(parts[0])
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	appRoleID, err := strconv.Atoi(parts[1])
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	resp := new(hc.FundingSourcePermissionsMappingListResponse)
+	err = client.GET(fmt.Sprintf("/v3/funding-source/%d/permission-mapping", fundingSourceID), resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	var diags diag.Diagnostics
+
+	found := false
+	for _, mapping := range resp.Data {
+		if mapping.AppRoleID == appRoleID {
+			// Set lists to the state as provided
+			diags = append(diags, hc.SafeSet(d, "funding_source_id", fundingSourceID)...)
+			diags = append(diags, hc.SafeSet(d, "app_role_id", appRoleID)...)
+			diags = append(diags, hc.SafeSet(d, "user_groups_ids", mapping.UserGroupsIDs)...)
+			diags = append(diags, hc.SafeSet(d, "user_ids", mapping.UserIDs)...)
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		d.SetId("")
+	}
+
+	return diags
+}
+
+func resourceFundingSourcePermissionsMappingUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
```

Comment:
Curious did you test something like this:

```hcl
kion_funding_source_permission_mapping "test1" {
  app_role_id = 1
  user_ids = [1, 2, 3]
}
```
Apply, and then try to change:
```hcl
kion_funding_source_permission_mapping "test1" {
  app_role_id = 2
  user_ids = [1, 2, 3]
}
```

I'm curious if this example removes the app_role_id=1 mapping and replaces with app_role_id=2 mapping?  The code looks like it wouldn't remove app_role_id=1 since you don't ever use the terraform ID() in the update.  But to be fair, I didn't test this myself.

Reply:
Actually, I did not.  I'll test.

Reply:
Good call.  It did not correctly remove it or add it to the new mapping.  I'll fix this.

Reply:
This is fixed.